cpp/memilio/compartments/compartmentalmodel.h

```diff
@@ -198,9 +165,255 @@ struct CompartmentalModel {
+
+public:
+    using Base      = CompartmentalModel<Comp, Pop, Params>;
+    using FlowChart = TypeChart<Flows...>;
+    /**
+     * @brief Default constructor, forwarding args to Base constructor.
+     */
+    template <class... Args>
+    CompartmentalModel<Comp, Pop, Params, FlowChart>(Args... args)
+        : CompartmentalModel<Comp, Pop, Params>(args...)
+        , m_comp_factor(comp_factor())
+        , m_flow_index_dimensions(reduce_index(this->populations.size()))
+        , m_flow_values((this->populations.numel() / static_cast<size_t>(Comp::Count)) * FlowChart().size())
+    {
+    }
+
+    // Note: use get_flow_index when accessing flows
+    // Note: by convention, we compute incoming flows, thus entries in flows must be non-negative
+    virtual void get_flows(Eigen::Ref<const Eigen::VectorXd>, Eigen::Ref<const Eigen::VectorXd> /*y*/, double /*t*/,
+                           Eigen::Ref<Eigen::VectorXd> /*flows*/) const {};
+
+    /**
+     * @brief Compute the right-hand-side of the ODE dydt = f(y, t) from flow values.
+     *
+     * This function is generated at compile time depending on the template parameters Flows and Pop of the model.
+     *
+     * @param[in] flows The current flow values (as calculated by get_flows) as a flat array.
+     * @param[out] dydt A reference to the calculated output.
+     */
+    void get_derivatives(Eigen::Ref<const Eigen::VectorXd> flows, Eigen::Ref<Eigen::VectorXd> dydt) const
+    {
+        dydt.setZero();
+        for (size_t i = 0; i < this->populations.numel() / static_cast<size_t>(Comp::Count); i++) {
+            // see get_flow_index for more info on these index calculations.
+            // the Comp-index is added in get_rhs_impl; the flat PopIndex is
+            //   flat_pop_index = (i % m_comp_factor) + (i / m_comp_factor) * m_comp_factor * Comp::Count;
+            // mind the integer division.
+            get_rhs_impl<0>(flows, dydt,
+                            i + (i / m_comp_factor) * m_comp_factor * (static_cast<size_t>(Comp::Count) - 1),
+                            i * FlowChart().size());
+        }
+    }
+
+    /**
+     * @brief Compute the right-hand-side f(y, t) of the ODE and store it in dydt.
+     *
+     * This function uses get_flow(..., flows) and get_derivatives(flows, dydt) to provide the
+     * same interface as a CompartmentalModel without flows.
+     *
+     * @param[in] pop The current population of the model as a flat array.
+     * @param[in] y The current state of the model as a flat array.
+     * @param[in] t The current time.
+     * @param[out] dydt A reference to the calculated output.
+     */
+    void get_derivatives(Eigen::Ref<const Eigen::VectorXd> pop, Eigen::Ref<const Eigen::VectorXd> y, double t,
+                         Eigen::Ref<Eigen::VectorXd> dydt) const override final
+    {
+        m_flow_values.setZero();
+        get_flows(pop, y, t, m_flow_values);
+        get_derivatives(m_flow_values, dydt);
+    }
```

Comment:
In this specific function, we might be able to use dydt, but in general there can be more flows than entries in dydt. When we want every flow value, we cannot use dydt without resizing, which I am almost certain will cause issues elsewhere.

I like the idea of the `flow` or `add_flow` function, but either way we have to distinguish between a) simply storing each flow value computed by get_flows or b) aggregating the values into dydt (i.e. subtracting value from Source, adding to Dest).

Not sure how to determine which case we need, as I would prefer not to overload or add a template to get_flows. We could use an object with (a reference to) some vector (e.g. dydt) and a *virtual* (add_)flows member, or maybe a bool to determine which function to use.

Reply:
Not sure I see the problem. If the add_flow function accumulates the flows in dydt like get_rhs_impl does right now, the number of flows don't matter. Why do we need to distinguish between storing and aggregating? Is get_flows used for anything else than here?

Reply:
We use get_flows to specifically calculate the flows between compartments in SimulationFlows. Aggregating makes this in general impossible, which is why we started this branch in the first place.

 This get_derivatives implementation allows us to compute the compartments instead, mainly so that the model is backwards compatible.